pokemon-service/src/test/java/cz/muni/fi/pa165/pokemon/league/participation/manager/service/GymServiceTest.java

```diff
@@ -0,0 +1,376 @@
+        when(gymDAO.findGymById(gymBrno.getId())).thenReturn(gymBrno);
+        when(gymDAO.getAllGyms()).thenReturn(Stream.of(gymPraha, gymBrno)
+                .collect(Collectors.toList()));
+
+    }
+
+    @Test
+    public void testFindById() {
+        assertThat(gymService.findGymById(gymPraha.getId()))
+                .isEqualToComparingFieldByField(gymPraha);
+    }
+
+    @Test
+    public void testFindByIdWithException() {
+        doThrow(dae).when(gymDAO).findGymById(errorneousGym.getId());
+
+        assertThatExceptionOfType(DataAccessException.class)
+                .as("Testing DAO exception")
+                .isThrownBy(() -> gymService.findGymById(errorneousGym.getId()));
+    }
+
+    @Test
+    public void testFindByNonExistingId() {
+        assertThat(gymService.findGymById(Long.MIN_VALUE)).isNull();
+    }
+
+    @Test
+    public void testUpdateGymLocation() {
+        String newLocation = "New Location";
+        gymService.updateGymLocation(gymPraha, newLocation);
+
+        assertThat(gymService.findGymById(gymPraha.getId()).getLocation())
+                .isEqualTo(newLocation);
+    }
+
+    @Test
+    public void testUpdateGymLocationWithException() {
+        doThrow(dae).when(gymDAO).updateGym(errorneousGym);
+
+        String newLocation = "New Location";
+
+        assertThatExceptionOfType(DataAccessException.class)
+                .as("Testing DAO exception")
+                .isThrownBy(() -> gymService.updateGymLocation(errorneousGym, newLocation));
+    }
+
+    @Test
+    public void testChangeGymTypeByOwnTrainer() {
+        PokemonType newType = PokemonType.STEEL;
+        boolean result = gymService.changeGymType(gymPraha, gymPraha.getGymLeader(), newType);
+        assertThat(result).isTrue();
+        assertThat(gymService.findGymById(gymPraha.getId()).getType())
+                .isEqualTo(newType);
+    }
+
+    @Test
+    public void testChangeGymTypeByOtherTrainer() {
+        PokemonType originalType = gymPraha.getType();
+        PokemonType newType = PokemonType.STEEL;
+        boolean result = gymService.changeGymType(gymPraha, otherTrainer, newType);
+        assertThat(result).isFalse();
+        assertThat(gymService.findGymById(gymPraha.getId()).getType())
+                .isEqualTo(originalType);
+    }
+
+    @Test
+    public void testChangeGymTypeWithException() {
+        doThrow(dae).when(gymDAO).updateGym(errorneousGym);
+
+        PokemonType newType = PokemonType.STEEL;
+
+        assertThatExceptionOfType(DataAccessException.class)
+                .as("Testing DAO exception")
+                .isThrownBy(() -> gymService.changeGymType(
+                errorneousGym,
+                errorneousGym.getGymLeader(),
+                newType));
+    }
+
+    @Test
+    public void testChangeGymLeader() throws EntityIsUsedException {
+        gymService.changeGymLeader(gymPraha, otherTrainer);
+
+        assertThat(gymService.findGymById(gymPraha.getId()).getGymLeader())
```

Comment:
findById is unnecessary, and you should check whether DAO's updateGym was called.